### DISCUSSION TITLE
Un-habits

### DISCUSSION THREAD

**Discussion by abrooks** [iSoron/uhabits#696] [Feature Requests]
I'm wondering if there's appetite for un-habits, i.e. tracking dessert, vices or other things to avoid. Today the only way to track that is to note, on a daily basis, that you _didn't_ do something all day where the converse is really what you want — the default is that you didn't do it unless you note as an event that you did.

It seems likely that un-habits would want the following attributes:
- Allowance (number of times that you can do something... i.e. number of desserts, drinks, etc. per week)
- Event count (i.e. *don't* more than 2x snacks per day; this is something I've wanted for positive habits as well: *do* stretch 3x day)
- Adjustments in reporting, streaks, etc. (since the meaning is different)

What do folks think?

**Comment by Kadweirdo** [comment]
I came here to suggest the same idea!

Some habits are okay to do once in a while, but you have to track them to make sure they don't get out of hand. Or you want to reduce their frequency until they're gone.

Most of these "un-habits" can be turned into a positive habit, which are easier (and more rewarding!) to track. But when the frequency of the bad habit is too low, this doesn't really work anymore.

Tracking positive habits should definitely be the recommended way as it is a lot more rewarding, but for some cases it doesn't make sense.

Some examples:

_Easy to turn into positives_
Snacking: Okay 2 days per week. So 5 days per week I can put 'Didn't snack', which feels great.
Alcohol: Same as snacking. 5 days per week I can put 'No alchol', which is nice.
Eating less meat: Positive equivalent is "Veggie day", so I put my goal to 3 veggie days per week.

_Difficult to turn into positives_
Takeout food: Home-cooking is our standard, but ordering food 1 or 2 times per month is okay. It is not rewarding or fun to put 28 days per month "Home-cooked a meal". We don't need to improve on this, we just want to keep track to make sure it doesn't get too much, and to easily check when was the last time.

Use alternative transportation: We don't need the car every day, and we try to use alternatives whenever possible. But it's stupid to mark every day "Didn't take the car", because A) we don't need it every day, and B) sometimes it's okay to take the car. The actual achievement is "Used alternative transportation", but you can't set a goal because that's different every month. We just want to track whenever we took the car, and try to keep that to a maximum of X times per month.

**Comment by hiqua** [comment]
Numeric habits are already implemented in the current alpha. I think we'd then just need a setting to say that a habit is negative, and adapt the colors, and possibly add a default value.

**Reply by ke352802081770314** [reply]
Hello @hiqua since you mention a path to negative habits now and the wont-fix #15 and #351 was quite some time ago, does that mean that the developer team has re-considered it? 
Numerical habits may be a game changer, since as illustrated they can't be inverted as was suggested (and I found plausible) for non-numerical goals.

**Comment by naught101** [comment]
This has already been wont-fixed: https://github.com/iSoron/uhabits/issues/351#issuecomment-355340227

**Comment by lashoun** [comment]
About inverse habits, it seems to me that a very easy way to implement it would be to make another category of habits where new days are checked positively by default,  ie, an auto-check feature. Depending on how the code is written, another category is maybe not even needed and it could be a setting that could be changed at any time, since it only operates on new days.
I believe this solution is very satisfactory because you would still have to reword the habit in a positive way (eg "Did I have a smoke-free day?") but you would only have to *manually uncheck* the day if you did smoke. This would remove the trouble of waiting until the end of the day or even the next morning to check the habit. Psychologically, this would really change the deal for me because if the day is unchecked by default, I feel like I do not *lose* anything by smoking, but if it was checked, I would have greater aversion to smoking since I would _lose_ the checkmark. [Loss aversion](https://www.wikiwand.com/en/Loss_aversion) is indeed a real bias, for me at least. Furthermore, there would be no other change needed on streaks, score, or anything else. @iSoron, I believe this does not go against your philosophy; what do you think?

**Reply by rootmeanclaire** [reply]
There has been an ongoing discussion about this under #1106

**Comment by juancamilo87** [comment]
I read the other discussion and I agree that when possible it would be better to change the unhabit into something positive. And this is easy for yes or kk questions.

But what for quantifiable things? What if you can have a certain amount of something but not over some limit? It would be nice to have that possibility, like eating too many sweets. And having a weekly goal of less than 3 desserts for example.

Also, I was thinking quantifiable habits could also use an upper bound for the goal. For example hours of exercise. For someone starting more is better, but too much is also bad. So it would be nice to make the goal something like between 4 and 8 hours a week, or something like that.

**Reply by hiqua** [reply]
Some code is already there regarding "at most" habits, so it's in the works, at least for a basic version of this.

**Comment by g2petter** [comment]
I'm new to the app, so it might be that my perspective changes over time, but here's my rationale for why I would like simple negative habits:

I want a more conscious relationship to my bad habits, so instead of asking "Did you avoid the cake in the break room?" at the end of the day, hours after the bad choice was made, I want to be able to flag the habit as broken as close to making the bad choice as possible.

I've done this with a pen and paper calendar once, and it was quite effective at breaking bad habits since the act of tracking the bad habit became a habit in itself.

**Reply by hiqua** [reply]
That's covered by having default values which is being implemented in https://github.com/iSoron/uhabits/pull/1106

**Reply by ke352802081770314** [reply]
If you activate the global setting "show question mark for unfilled days" (wording may be a bit different, I am shown my language version so I am guessing the English), you can do what you are suggesting: it starts with "?", tap for tick and tap again for x (fail) immediately when you lapsed (note if you have activated skip, you have to tap a third time to get to "fail"), and at the end of the day you can set your ? into a tick (success) if you didn't fail.

**Reply by herrbrixter** [reply]
> That's covered by having default values which is being implemented in #1106

#1106 wasn't merged though. Currently, there is no solution for negative habits.

**Comment by devingfx** [comment]
Hi!
Another example:
Be able to count cigarette during the day, with maybe an increment button widget, with a goal of "less than 10 cigarette by day".
It could be achieved by adding a "less than" / "greater than" select on the counted habit... Then the habit is validated when the number is less than the goal. Instead of being validated like today only if greater than the goal...

**Reply by hiqua** [reply]
There are "at most" habits implement in the release coming out soon already.

**Reply by nilsherzig** [reply]
"at most" habits currently start with a 100% rating, which in my opinion doesn't work well with the graph or little circles on the start page. Would it be possible to also have them start with a 0% rating?

**Comment by chendy** [comment]
Another app has options of 
'at least' vs 'less than' vs 'exactly' vs 'any value'.
Seems simple enough, and think addresses all possible use cases

**Reply by hiqua** [reply]
This already exists on the `dev` version which should be released soon.

P.S.: please don't advertise non-FOSS apps here.

**Comment by manoelpqueiroz** [comment]
I would like to give my two cents here, as this is a discussion from a long time now.

While I believe the app already implements negative numeric habits via the "at most" option, I think we could benefit from implementing them in yes/no habits.

For a reference, I really like the notion behind this use case in [ToDon't](https://github.com/Crazy-Marvin/ToDont). I believe that for Loop, I would incline to say it's achievable through aesthetics:

- By setting a habit to be "negative", the default behaviour of the tracker should be inverted;
- This means that every new day, the default is to mark a :heavy_check_mark: with gray indicating the habit has failed (i.e., unless the user specifically indicates they have not done that, it should count like the default grey :x: for normal habits);
- Conversely, if the user "checks" they have not done the habit, the :x: should be coloured;
- Score for the habits wouldn't need to change, as it should start from 0 and build up for every day the user checks a "didn't do".

I haven't worked with front-end, but at least it seems less complicated than I would normally assess. What do you guys think?

**Reply by herrbrixter** [reply]
This is the best solution. Just reverse the mechanics of a positive habit.

It's sad that we have to use a workaround solution using 'at most" without an elegant solution. This is third most requested feature.

**Comment by herrbrixter** [comment]
@iSoron Just reverse the mechanics of a positive habit. The question marks need to be changed to distinguish positive from negative habits.

Positive habits:

- Default value is Failed: gray ⛌ (counts as Failed.)
- Completed Successfully: colored ✔️ (counts as Success.)
- Explicitly marked as Failed: colored ❌ (counts as Failed.)

Negative habits:

- Default value is Success: gray ✓ (counts as Success.)
- Failed the bad habit: colored ❌ (counts as Failed.)
- Explicitly marked as Success: colored ✔️ (counts as Success.)

By replacing question marks (?, ?, ?, ?, ?) with gray checks and crosses (✓, ✓, ⛌, ⛌, ✓) it is easier to tell which are positive and negative habits.
